File: atcoder.jp/PyPy3 (7.3.0)/abc277/abc277_c.py
from collections import defaultdict, deque
from typing import List, Dict

Node  = int
Graph = Dict[Node, List[Node]]
Depth = Dict[Node, int]
Queue = List[Node]


def bfs(graph: Graph, n: Node) -> Depth:
    depth: Depth = defaultdict(lambda : -1)
    depth[n] = 0
    queue = deque()
    queue.append(n)

    while queue:
        u = queue.popleft()
        for v in graph[u]:
            if depth[v] == -1:
                queue.append(v)
                depth[v] = depth[u] + 1

    return depth

N = int(input())

graph = defaultdict(list)

for _ in range(N):
    A, B = map(int, input().split())
    graph[A].append(B)
    graph[B].append(A)


depth = bfs(graph, 1)

ans = max(depth.keys())

print(ans)

File: atcoder.jp/Rust (1.42.0)/abc277/abc277_c.rs
use proconio::{input, fastout};
use std::collections::{HashMap, VecDeque};

fn bfs<T>(graph: &HashMap<T, Vec<T>>, start: T) -> HashMap<T, usize>
where
    T: Eq + std::hash::Hash + Copy,
{
    let mut queue = VecDeque::new();
    queue.push_back(start);
    let mut depth = HashMap::new();
    depth.insert(start, 0);
    while let Some(u) = queue.pop_front() {
        for &v in graph.get(&u).unwrap() {
            if !depth.contains_key(&v) {
                depth.insert(v, depth[&u] + 1);
                queue.push_back(v);
            }
        }
    }
    depth
}

#[fastout]
fn main() {
    input! {
        n: usize,
        ab: [(usize, usize); n],
    }

    let graph = ab.into_iter().fold(HashMap::new(), |mut graph, (a, b)| {
        graph.entry(a).or_insert(vec![]).push(b);
        graph.entry(b).or_insert(vec![]).push(a);
        graph
    });

    if !graph.contains_key(&1) {
        println!("1");
        return;
    }

    let depth = bfs(&graph, 1);

    let ans = depth.keys().max().unwrap();

    println!("{}", ans);
}
